### PR TITLE
PXB 3038 Merge MySQL 8.0.33 in Xtrbackup

### PR DESCRIPTION
**Pull request by rahulmalik87**
Merged 8.0.33 in PXB

Also 
In Q2 2023 Oracle released MySQL Server 5.7.42 and 8.0.33 and completely
re-wrote git commit history in their git repo.
This "fixing" null merge is supposed to bring new Oracle mysql-server commit
hashes (up to mysql-server-8.0.32) into existing PXB  8.0 branch.

### CODE REVIEW

File: storage/innobase/srv/srv0start.cc [hunk 1]
```diff
@@ -2037,15 +2021,11 @@ dberr_t srv_start(bool create_new_db) {
...
-<<<<<<< HEAD
-    err = recv_recovery_from_checkpoint_start(*log_sys, flushed_lsn, to_lsn);
-||||||| ce0de82d3aa
-    err = recv_recovery_from_checkpoint_start(*log_sys, flushed_lsn);
-=======
     ut_a(log_sys->m_format <= Log_format::CURRENT);
 
     const bool log_upgrade = log_sys->m_format < Log_format::CURRENT;
 
+    // 8033 merge recreate redo
```
Comment (altmannmarcelo) [inline storage/innobase/srv/srv0start.cc]: what does this means?

File: storage/innobase/srv/srv0start.cc [hunk 2]
```diff
@@ -2037,15 +2021,11 @@ dberr_t srv_start(bool create_new_db) {
...
     if (log_upgrade) {
```
Comment (altmannmarcelo) [inline storage/innobase/srv/srv0start.cc]: When running --prepare on an old backup (pre 8.0.30) does we fall into this if?
can you test it under gdb and paste the results here ?

Comment (altmannmarcelo) [reply storage/innobase/srv/srv0start.cc]: If I am not mistaken, we adjust xtrabackup_logfile format to be Log_format::CURRENT before renaming it to ib_redo0

File: storage/innobase/srv/srv0start.cc [hunk 3]
```diff
@@ -2229,309 +2203,14 @@ dberr_t srv_start(bool create_new_db) {
...
-    DBUG_EXECUTE_IF("log_force_resize", log_downsize_requested = true;);
-
-    const bool need_to_recreate_log_files =
 #ifdef XTRABACKUP
         /* we have to recreate always in PXB because the ib_redo0 file
         is always full and we cannot reuse it. No circular buffer
         logic from 8.0.30. Recreate this #ib_redo0 file and it will be
         used by transaction rollbacks */
-        (!srv_apply_log_only && !srv_read_only_mode) ||
+    // 8033 merge todo recreate redo
```
Comment (altmannmarcelo) [inline storage/innobase/srv/srv0start.cc]: this seems like a TODO missing to be implemented.

File: storage/innobase/xtrabackup/src/xbcloud/CMakeLists.txt [hunk 4]
```diff
@@ -61,6 +60,8 @@ TARGET_LINK_LIBRARIES(xbcloud
   ${GCRYPT_LIBS}
   ${LIBEV_LIBRARIES}
   ${CURL_LIBRARIES}
+  extra::rapidjson
```
Comment (altmannmarcelo) [inline storage/innobase/xtrabackup/src/xbcloud/CMakeLists.txt]: Why rapidjson is needed here ? Seems like it was not required before I remember we only use it on xtrabackup to deal with keyrings, and that is all done on keyring_component.cc

File: storage/innobase/xtrabackup/src/xbcloud/CMakeLists.txt [hunk 5]
```diff
@@ -61,6 +60,8 @@ TARGET_LINK_LIBRARIES(xbcloud
...
+  ext::curl
```
Comment (altmannmarcelo) [inline storage/innobase/xtrabackup/src/xbcloud/CMakeLists.txt]: what is the difference between `ext::curl`  and ` ${CURL_LIBRARIES}` linked at line 62?

File: storage/innobase/xtrabackup/src/xbcloud/CMakeLists.txt [hunk 6]
```diff
@@ -89,6 +90,8 @@ IF (WITH_UNIT_TESTS AND CMAKE_VERSION VERSION_GREATER 3.22.1)
       ${GTEST_BOTH_LIBRARIES}
       GTest::gtest
       GTest::gmock
+      extra::rapidjson
+      ext::curl
```
Comment (altmannmarcelo) [inline storage/innobase/xtrabackup/src/xbcloud/CMakeLists.txt]: same as above

File: storage/innobase/xtrabackup/src/xtrabackup.cc [hunk 7]
```diff
@@ -7232,11 +7208,6 @@ static void xtrabackup_prepare_func(int argc, char **argv) {
 
   srv_apply_log_only = (bool)xtrabackup_apply_log_only;
 
-  /* increase IO threads */
```
Comment (altmannmarcelo) [inline storage/innobase/xtrabackup/src/xtrabackup.cc]: This seemed like a performance optimization, shall we keep read/write io threads to 4 unconditionally or find another metric to decide if we need to increase?

File: storage/innobase/srv/srv0start.cc [hunk 8]
```diff
@@ -2551,6 +2551,18 @@ void srv_start_threads() {
     purge_sys->state = PURGE_STATE_DISABLED;
     return;
   }
+#ifdef XTRABACKUP
```
Comment (altmannmarcelo) [inline storage/innobase/srv/srv0start.cc]: My understanding of https://github.com/percona/percona-xtrabackup/commit/1972123c5d7a11f97de11a292a317410ee9d2b29 is that we moved the rollback to after DDL recovery, which is done as part of --prepare.

What problem where you seeing to revert this change?

File: storage/innobase/srv/srv0start.cc [hunk 9]
```diff
@@ -2928,7 +2583,9 @@ void srv_start_threads() {
 }
 
 void srv_start_threads_after_ddl_recovery() {
-  if (srv_force_recovery < SRV_FORCE_NO_TRX_UNDO && trx_sys_need_rollback()) {
+  if (IF_XB(!srv_apply_log_only &&)
```
Comment (altmannmarcelo) [inline storage/innobase/srv/srv0start.cc]: We could potentially add the same IF_XB(!srv_apply_log_only) at https://github.com/rahulmalik87/percona-xtrabackup/blob/pxb-8.0-3038/storage/innobase/trx/trx0trx.cc#L810, no reason on populating the list of transactions to rollback if we are not going to roll them back.

File: storage/innobase/xtrabackup/src/xtrabackup.cc [hunk 10]
```diff
@@ -6849,17 +6849,17 @@ static bool xb_export_cfg_write(
...
-    const dict_table_t *table, FILE *file) {
+    const Encryption_metadata &encryption_metadata, FILE *file) {
   byte key_size[sizeof(uint32_t)];
   byte row[Encryption::KEY_LEN * 3];
   byte *ptr = row;
   byte *transfer_key = ptr;
   lint elen;
-
-  ut_ad(table->encryption_key != NULL && table->encryption_iv != NULL);
+  ut_a(encryption_metadata.can_encrypt());
```
Comment (altmannmarcelo) [inline storage/innobase/xtrabackup/src/xtrabackup.cc]: shall we keep:

1. ut_ad
2. testing encryption_metadata.m_key and encryption_metadata.m_iv are not null ? can_ecnrypt will check if m_type is not Encryption::NONE but will not ensure m_key m_iv are populated.

File: storage/innobase/xtrabackup/src/redo_log.cc [hunk 11]
```diff
@@ -731,29 +731,23 @@ bool Redo_Log_Writer::close_logfile() {
...
-    ut_a(len == dst_len);
-  }
 
+    if (!encryption.encrypt_log(buf, len, scratch_buf)) {
+      xb::error() << "Failed to encrypt redo log";
+      return false;
+    }
+    write_buf = scratch_buf;
+  }
   if (ds_write(log_file, write_buf, len)) {
```
Comment (altmannmarcelo) [inline storage/innobase/xtrabackup/src/redo_log.cc]: Do we need write_buf at all here?
scratch_buf seems to have what we need.

File: storage/innobase/xtrabackup/src/CMakeLists.txt [hunk 12]
```diff
@@ -119,13 +119,12 @@ TARGET_LINK_LIBRARIES(xtrabackup
   innobase
   sql_dd
   sql_gis
-  sql_main
```
Comment (altmannmarcelo) [inline storage/innobase/xtrabackup/src/CMakeLists.txt]: isn't sql_main required anymore ?

**Review by altmannmarcelo — CHANGES_REQUESTED** [review]
@rahulmalik87 thanks for working on this.

Please convert this to ready-to-review PR adjusting the title and description.

First round of review comments.

File: storage/innobase/xtrabackup/src/xbcloud/CMakeLists.txt (hunk 4 above)
Comment (rahulmalik87) [reply storage/innobase/xtrabackup/src/xbcloud/CMakeLists.txt]: Used in swift.cc
/Users/rahulmalik/MySQL/src/x38/storage/innobase/xtrabackup/src/xbcloud/swift.cc:30:10: fatal error: 'rapidjson/document.h' file not found
#include <rapidjson/document.h>

File: storage/innobase/xtrabackup/src/xbcloud/CMakeLists.txt (hunk 5 above)
Comment (rahulmalik87) [reply storage/innobase/xtrabackup/src/xbcloud/CMakeLists.txt]: removing the CURL_LIBRARIES. we just need ext::curl.

File: storage/innobase/xtrabackup/src/CMakeLists.txt (hunk 12 above)
Comment (rahulmalik87) [reply storage/innobase/xtrabackup/src/CMakeLists.txt]: It was duplicated.

File: storage/innobase/xtrabackup/src/xtrabackup.cc (hunk 7 above)
Comment (rahulmalik87) [reply storage/innobase/xtrabackup/src/xtrabackup.cc]: the default value is 4 or the user can provide innobase_read_io_threads.

File: storage/innobase/srv/srv0start.cc (hunk 8 above)
Comment (rahulmalik87) [reply storage/innobase/srv/srv0start.cc]: rollback to after DDL recovery is part of [addd56b38](https://github.com/percona/percona-xtrabackup/commit/addd56b38).
Server calls rollback from `srv_start_threads_after_ddl_recovery` which is called the innobase code which we do not call

File: storage/innobase/srv/srv0start.cc (hunk 9 above)
Comment (rahulmalik87) [reply storage/innobase/srv/srv0start.cc]: We do not populate the list of transactions to rollback if we are not going to rollback. It is handled in function [`trx_rseg_add_rollback_segments`.](https://github.com/rahulmalik87/percona-xtrabackup/blob/809cf460e7107787ebe250d7d18b48e14811738b/storage/innobase/trx/trx0rseg.cc#L876)
If you want I an add a debug assert or something?

File: storage/innobase/xtrabackup/src/xtrabackup.cc (hunk 10 above)
Comment (rahulmalik87) [reply storage/innobase/xtrabackup/src/xtrabackup.cc]: Fixed

File: storage/innobase/xtrabackup/src/redo_log.cc (hunk 11 above)
Comment (rahulmalik87) [reply storage/innobase/xtrabackup/src/redo_log.cc]: if redo log are not encrypted we would just write write_buf.

File: storage/innobase/srv/srv0start.cc (hunk 3 above)
Comment (rahulmalik87) [reply storage/innobase/srv/srv0start.cc]: PXB is now recreating the redo before rollback of transaction.

File: storage/innobase/srv/srv0start.cc (hunk 1 above)
Comment (rahulmalik87) [reply storage/innobase/srv/srv0start.cc]: removed,  it was todo

File: storage/innobase/srv/srv0start.cc (hunk 2 above)
Comment (rahulmalik87) [reply storage/innobase/srv/srv0start.cc]: 8.0.28 . 
`Thread 1 "xtrabackup" hit Breakpoint 1, srv_start (create_new_db=false,
    to_lsn=18152822) at ../storage/innobase/srv/srv0start.cc:2034
2034        if (log_upgrade) {
(gdb) bt
#0  srv_start (create_new_db=false, to_lsn=18152822) at ../storage/innobase/srv/srv0start.cc:2034
#1  0x00005555587cf06f in innodb_init (init_dd=true, for_apply_log=true) at ../storage/innobase/xtrabackup/src/xtrabackup.cc:2674
#2  0x00005555587e00b3 in xtrabackup_prepare_func (argc=7, argv=0x55555e1d19e0) at ../storage/innobase/xtrabackup/src/xtrabackup.cc:7196
#3  0x00005555587e3f8e in main (argc=8, argv=0x7fffffffd4e8) at ../storage/innobase/xtrabackup/src/xtrabackup.cc:8255
(gdb) p log_upgrade
$1 = false`

**Comment by rahulmalik87** [comment]
Jenkin results. https://pxb.cd.percona.com/job/percona-xtrabackup-8.0-test-param/314/testReport/

All failures are due to [Duplicate](https://jira.percona.com/browse/PXB-2865) keys issue . PXB-1679.sh is known [failure](https://jira.percona.com/browse/PXB-2948) due to concurrent DDL.

**Comment by altmannmarcelo** [comment]
Taking over at #1452
